lib/Conch/Controller/DeviceReport.pm

```diff
@@ -28,36 +28,48 @@ sub process ($c) {
 	my $device_report = $c->validate_input('DeviceReport');
 	return if not $device_report;
 	my $raw_report = $c->req->body;
+	my $schema = $c->schema;
 
-	my $maybe_hw;
+	my $hw;
+	if ($device_report->{device_type} && $device_report->{device_type} eq "switch") {
+		$hw = $schema->resultset('HardwareProduct')->find({
+			name => $device_report->{product_name}
+		});
 
-	if ( $device_report->{device_type}
-		&& $device_report->{device_type} eq "switch" )
-	{
-		$maybe_hw = Conch::Model::HardwareProduct->lookup_by_name(
-			$device_report->{product_name}
-		);
-		return $c->status(409, {
+		$hw or return $c->render(status => 409, json => {
 			error => "Hardware product name '".$device_report->{product_name}."' does not exist"
-		}) unless ($maybe_hw);
+		});
 
 	} else {
-		$maybe_hw = Conch::Model::HardwareProduct->lookup_by_sku(
-			$device_report->{sku}
-		);
+		$hw = $schema->resultset('HardwareProduct')->find({
+			sku => $device_report->{sku}
+		});
+
+		if(not $hw) {
+			$c->log->debug("Could not find hardware product by SKU, falling back to legacy_product_name");
+		    $hw = $schema->resultset('HardwareProduct')->find({
+				legacy_product_name => $device_report->{product_name}
+			});
+
+			$hw or return $c->render(status => 409, json => {
```

Comment:
This should be a 404 no?

Reply:
I'd vote not because the user is doing a `POST /device/:id` and a 404 might be interpreted to mean that the device doesn't exist. This is validating data inside the report itself.

Reply:
Fair enough